os/net/lwip/test/unit/mbox/mbox_shared.c

```diff
@@ -0,0 +1,182 @@
+  receive a signal.
+  6) Unexpected behavior will happen.
+  */
+
+  /*  send signal to thread a to call sys_arch_mbox_fetch */
+  printf("[COND] run task A \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_post(g_signal_ctoa);
+  /*  wait event that task a wait event with mbox->mail because it's empty */
+  printf("[COND] wait until task A fetch event \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_wait(g_signal_atoc);
+  printf("[COND] task A wait not empty signal \t%s:%d\n",  __FUNCTION__, __LINE__);
+  /*  let task b to post data until g_tcpmbox is full */
+  printf("[COND] run task B to post data \t%s:%d\n",  __FUNCTION__, __LINE__);
+  sem_post(g_signal_ctob);
+
+  return NULL;
+}
+
+static void *a_thread(void *arg) {
+  uint32_t recv_cnt = 0;
+  printf("[taskA] wait control signal \t%s:%d\n", __FUNCTION__, __LINE__);
+  sem_wait(g_signal_ctoa);
+  while (recv_cnt < QUEUE_MSG_CNT) {
+    lwip_msg_s *msg = NULL;
+    printf("[taskA] -->sys_arch_mbox_fetch \t%s:%d\n", __FUNCTION__,
+           __LINE__);
+    uint32_t res = sys_arch_mbox_fetch(&g_tcpmbox, (void **)&msg, 0);
+    if (res != 0) {
+      assert(0);
+    }
+    free(msg);
+    recv_cnt++;
+  }
+  return NULL;
+}
+
+static void *b_thread(void *arg) {
+  printf("[taskB] wait control signal  \t%s:%d\n", __FUNCTION__, __LINE__);
+  sem_wait(g_signal_ctob);
+  uint32_t send_cnt = 0;
+  while (send_cnt < QUEUE_MSG_CNT) {
+    lwip_msg_s *msg = (lwip_msg_s *)malloc(sizeof(lwip_msg_s));
+    if (!msg) {
+      assert(0);
+    }
+    msg->type = APP_MSG;
+    msg->data = send_cnt + 1000;
+    sys_mbox_post(&g_tcpmbox, (void *)msg);
+    send_cnt++;
+  }
+  return NULL;
+}
+
+static int _init_resource(void) {
+  ST_ASSERT_EQ(ERR_OK, sys_mbox_new(&g_tcpmbox, TCP_QUEUE_SIZE));
+  printf("[TEST] mbox created %p %d\n", &g_tcpmbox, TCP_QUEUE_SIZE);
+  g_signal_atoc = (sem_t *)malloc(sizeof(sem_t));
+  ST_ASSERT_NEQ(NULL, g_signal_atoc);
+  g_signal_ctoa = (sem_t *)malloc(sizeof(sem_t));
+  ST_ASSERT_NEQ(NULL, g_signal_ctoa);
+  g_signal_btoc = (sem_t *)malloc(sizeof(sem_t));
+  ST_ASSERT_NEQ(NULL, g_signal_btoc);
+  g_signal_ctob = (sem_t *)malloc(sizeof(sem_t));
+  ST_ASSERT_NEQ(NULL, g_signal_ctob);
+
+  ST_ASSERT_EQ(0, sem_init(g_signal_atoc, 0, 0));
+  ST_ASSERT_EQ(0, sem_init(g_signal_ctoa, 0, 0));
+  ST_ASSERT_EQ(0, sem_init(g_signal_btoc, 0, 0));
+  ST_ASSERT_EQ(0, sem_init(g_signal_ctob, 0, 0));
+}
+
+static int _deinit_resource(void) {
+  ST_ASSERT_EQ(0, sem_destroy(g_signal_atoc));
+  ST_ASSERT_EQ(0, sem_destroy(g_signal_ctoa));
+  ST_ASSERT_EQ(0, sem_destroy(g_signal_btoc));
+  ST_ASSERT_EQ(0, sem_destroy(g_signal_ctob));
+
+  free(g_signal_atoc);
+  free(g_signal_ctoa);
+  free(g_signal_btoc);
+  free(g_signal_ctob);
+}
+
+static void _print_result(void) { printf("[TEST] test done\n"); }
+
+int shared_test(void) {
+  pthread_t contid, taid, tbid;
+
+  (void)_init_resource();
+
+  ST_ASSERT_EQ(0, pthread_create(&contid, NULL, control_thread, NULL));
```

Comment:
`pthread_create` is also user side API, so please use `kthread_create`.

Reply:
Thanks. I'll modify it with wrapper.